src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,349 @@
+        }
+        for (let i = 0; i < a.thumbnails.length; i++) {
+          if (a.thumbnails[i].start !== b.thumbnails[i].start ||
+              a.thumbnails[i].duration !== b.thumbnails[i].duration ||
+              a.thumbnails[i].mediaURL !== b.thumbnails[i].mediaURL) {
+            return false;
+          }
+        }
+        return true;
+      }),
+      switchMap(({ thumbnails, time }) => {
+        return videoSourceInfos$.pipe(
+          mergeMap((videoSourceBuffer) => {
+            return castToObservable(this.removeBuffers(videoSourceBuffer, time)).pipe(
+              mergeMap(() => {
+                if (!this._thumbnailTrack) {
+                  throw new Error(
+                    "VideoThumbnailLoaderError: No thumbnail track given.");
+                }
+
+                return getSegmentData(thumbnails, videoElement).pipe(
+                  mergeMap((datas) => {
+                    if (datas) {
+                      const appendBuffers$ = combineLatest(
+                        datas.map((data) => {
+                          return videoSourceBuffer
+                            .appendBuffer({
+                              segment: data,
+                              initSegment: null,
+                              codec: this._thumbnailTrack.codec,
+                              timestampOffset: 0,
+                            });
+                        })
+                      );
+                      return appendBuffers$.pipe(
+                          mergeMap(() => {
+                            thumbnails.forEach((t) => {
+                              this._bufferedDataRanges.push({
+                                start: t.start,
+                                end: t.start + t.duration,
+                              });
+                            });
+                            this._thumbnailVideoElement.currentTime = time;
+                            return observableOf(null);
+                          }),
+                          catchError((err) => {
+                            throw new Error(
+                              "VideoThumbnailLoaderError: Couldn't append buffer :" +
+                              err.message || err
+                            );
+                          })
+                        );
+                    }
+                    return observableOf(null);
+                  })
+                );
+              }));
+          })
+        );
+      }),
+      catchError((err) => {
+        this.dispose();
+        throw err;
+      })
+    ).subscribe();
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time, this._thumbnailVideoElement.currentTime
+   * @returns {Promise}
+   */
+  setTime(time: number) {
+    this._setTime$.next(time);
```

Comment:
I had an idea. Wouldn't it be better for the application if that API returned a Promise?

Which is:
  - resolved when the corresponding data has been correctly pushed
  - rejected when either on error or canceled

That way, they could decide to only display the data when it was received or to actually know when the right data is displayable